include/flucoma/clients/nrt/DataSetClient.hpp

```diff
@@ -337,22 +337,132 @@ class DataSetClient : public FluidBaseClient,
+
+  void setParams(ParamSetViewType& p) { mParams = p; }
+
+  template <size_t N>
+  auto& get() const
+  {
+    return mParams.get().template get<N>();
+  }
+
+  static constexpr auto& getParameterDescriptors() { return DataSetReadParams; }
+
+  DataSetRead(ParamSetViewType& p, FluidContext& c)
+      : mParams(p)
+  {
+    controlChannelsIn(1);
+    controlChannelsOut({1, 1});
+  }
+
+  index latency() const { return 0; }
+
+  template <typename T>
+  void process(std::vector<FluidTensorView<T, 1>>& input,
+               std::vector<FluidTensorView<T, 1>>& output, FluidContext& c)
+  {
+    if (input[0](0) > 0)
+    {
+      auto inputDSpointer = get<kDataSet>().get().lock();
+      if (!inputDSpointer)
+        return; // c.reportError("FluidDataSet RT Query: No FluidDataSet
+                // found");
+
+      index k = get<kNumNeighbors>();
+      if (k > inputDSpointer->size() || k < 0)
+        return; // c.reportError("FluidDataSet RT Query has wrong k size");
+
+      index dims = inputDSpointer->dims();
+
+      InOutBuffersCheck bufCheck(dims);
+      if (!bufCheck.checkInputs(get<kInputBuffer>().get(),
+                                get<kOutputBuffer>().get()))
+        return; // c.reportError("FluidDataSet RT Query i/o buffers are
+                // unavailable");
+
+      auto lookupDSpointer = get<kLookupDataSet>().get().lock();
+
+      index pointSize = lookupDSpointer ? lookupDSpointer->dims().value() : 1;
+
+      auto outBuf = BufferAdaptor::Access(get<kOutputBuffer>().get());
+      mNumValidKs = outBuf.samps(0).size() / pointSize;
+      if (mNumValidKs <= 0)
+      {
+        output[0](0) = 0;
+        return;
+      }
+      index outputSize = mNumValidKs * pointSize;
+
+      RealVector point(dims, c.allocator());
+      point <<= BufferAdaptor::ReadAccess(get<kInputBuffer>().get())
+                    .samps(0, dims, 0);
+
+      auto inputdata = inputDSpointer->getDataSet().getData();
+      auto distances = sortedDistances(point, inputdata, c.allocator());
+      auto outSamps = outBuf.samps(0);
+
+      if (lookupDSpointer)
+      {
+        auto lookupDS = lookupDSpointer->getDataSet();
+        auto inputDSids = inputDSpointer->getDataSet().getIds();
+        std::for_each_n(distances.begin(), mNumValidKs,
+                        [pointSize, &outSamps, &lookupDS, &inputDSids,
+                         n = 0](auto& p) mutable {
+                          if (auto point = lookupDS.get(inputDSids[p.first]);
+                              point.data() != nullptr)
+                          {
+                            outSamps(Slice(n, pointSize)) <<= point;
+                          }
+                          n += pointSize;
+                        });
+      }
+      else
+      {
+        std::for_each_n(distances.begin(), mNumValidKs,
+                        [&outSamps, n = 0](auto& p) mutable {
+                          outSamps[n++] = std::sqrt(p.second);
+                        });
```

Comment:
This can be a transform

```suggestion 
        std::transform(distances.begin(), distances.begin() + numPoints,
                       outSamps.begin(),
                       [](auto p) { return std::sqrt(p.second); });
```